core/runtime/execute_engine.cpp

```
@@ -181,7 +183,270 @@ std::vector<at::Tensor> create_output_tensors(c10::intrusive_ptr<TRTEngine> comp
+
+    // Whether cudagraphs needs to record the graph on this pass
+    auto result = compiled_engine->runtime_states.set_runtime_states(
+        cudagraphs_enabled, compiled_engine->use_pre_allocated_outputs, shape_changed);
+
+    bool need_cudagraphs_record = std::get<0>(result);
+    bool can_use_pre_allocated_outputs = std::get<1>(result);
+    bool need_cudagraphs_reset = std::get<2>(result);
+
+    if (need_cudagraphs_reset) {
+      compiled_engine->cudagraph.reset();
+    }
+
+    std::vector<at::Tensor> outputs(compiled_engine->num_io.second);
+
+    // Intialize inputs and outputs to be available throughout the succeeding scopes
+    { // Input Setup
+      std::unique_ptr<torch::autograd::profiler::RecordProfile> input_profiler_guard;
+      if (compiled_engine->profile_execution) {
+        input_profiler_guard =
+            std::make_unique<torch::autograd::profiler::RecordProfile>(compiled_engine->input_profile_path);
+      }
+
+      setup_input_tensors(inputs, compiled_engine, cudagraphs_enabled, need_cudagraphs_record);
+      // Check if input shapes can be inferred.
+      int32_t const io_size{compiled_engine->cuda_engine->getNbIOTensors()};
+      std::vector<char const*> names(io_size);
+      int32_t const nbNames = compiled_engine->exec_ctx->inferShapes(names.size(), names.data());
+      TORCHTRT_CHECK(
+          nbNames == 0,
+          "The shapes of the inputs: "
+              << names
+              << " cannot be inferred. This could happen if the input tensor addresses/shapes haven't been configured correctly");
+    }
+
+    { // Output Setup
+      std::unique_ptr<torch::autograd::profiler::RecordProfile> output_profiler_guard;
+      if (compiled_engine->profile_execution) {
+        output_profiler_guard =
+            std::make_unique<torch::autograd::profiler::RecordProfile>(compiled_engine->output_profile_path);
+      }
+      if (can_use_pre_allocated_outputs) {
```

What are your thoughts on if pre-allocated outputs feature is needed with this OA ? Do they complement each other or not related ? cc: @keehyuna

---

`use_pre_allocated_outputs` will be silently ignored for now because it is expected to use in CG mode. If OA is enabled, there's no code calling `use_pre_allocated_outputs`.